packages/async-flow/src/replay-membrane.js

```diff
@@ -0,0 +1,393 @@
+        if (guestPromiseMap.get(promise) !== 'settled') {
+          /** @type {LogEntry} */
+          const entry = harden(['doFulfill', hVow, hostFulfillment]);
+          log.pushEntry(entry);
+          interpretOne(topDispatch, entry);
+        }
+      },
+      async hostReason => {
+        await log.promiseReplayDone();
+        if (guestPromiseMap.get(promise) !== 'settled') {
+          /** @type {LogEntry} */
+          const entry = harden(['doReject', hVow, hostReason]);
+          log.pushEntry(entry);
+          interpretOne(topDispatch, entry);
+        }
+      },
+    );
+    return promise;
+  };
+  harden(makeGuestForHostVow);
+
+  const { guestToHost, hostToGuest } = makeConvertKit(
+    bijection,
+    makeGuestForHostRemotable,
+    makeGuestForHostVow,
+  );
+
+  // /////////////////////////////// Interpreter ///////////////////////////////
+
+  /**
+   * These are the only ones that are driven from the interpreter loop
+   */
+  const topDispatch = harden({
+    doFulfill,
+    doReject,
+    // doCall, // unimplemented in the current plan
+  });
+
+  /**
+   * These are the only ones that are driven from the interpreter loop
+   */
+  const nestDispatch = harden({
+    // doCall, // unimplemented in the current plan
+    doReturn,
+    doThrow,
+  });
+
+  const interpretOne = (dispatch, [op, ...args]) => {
+    try {
+      op in dispatch ||
+        // separate line so I can set a breakpoint
+        Fail`unexpected dispatch op: ${q(op)}`;
+      return dispatch[op](...args);
+    } catch (problem) {
+      throw panic(problem);
+    }
+  };
+
+  const logDo = (dispatch, entry) => {
+    log.pushEntry(entry);
+    return interpretOne(dispatch, entry);
+  };
+
+  const callStack = [];
+
+  let unnestFlag = false;
+
+  /**
+   * @param {number} callIndex
+   * @returns {Outcome | undefined}
+   */
+  const nestInterpreter = callIndex => {
+    callStack.push(callIndex);
+    while (log.isReplaying() && !stopped) {
+      const entry = log.nextEntry();
+      const optOutcome = interpretOne(nestDispatch, entry);
+      if (unnestFlag) {
+        optOutcome ||
+          // separate line so I can set a breakpoint
+          Fail`only unnest with an outcome: ${q(entry[0])}`;
+        unnestFlag = false;
+        return optOutcome;
+      }
+    }
+    unnestFlag = false;
+  };
+
+  /**
+   * @param {number} callIndex
+   */
+  const unnestInterpreter = callIndex => {
+    !stopped ||
+      Fail`This membrane stopped. Restart with new membrane ${replayMembrane}`;
+    callStack.length >= 1 ||
+      // separate line so I can set a breakpoint
+      Fail`Unmatched unnest: ${q(callIndex)}`;
+    const i = callStack.pop();
+    i === callIndex ||
+      // separate line so I can set a breakpoint
+      Fail`Unexpected unnest: ${q(callIndex)} vs ${q(i)}`;
+    unnestFlag = true;
+    if (callStack.length === 0) {
+      void E.when(undefined, wake);
+    }
+  };
+
+  let stopped = false;
```

Comment:
We probably should move this earlier in the file

Reply:
Done